### PR TITLE
Fixed #27854 -- Added system check for nonexistent directories in STATICFILES_DIRS setting.

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
ticket-27854

### CODE REVIEW

File: docs/ref/checks.txt [hunk 1]
```diff
@@ -869,3 +869,5 @@ configured:
   contain the :setting:`STATIC_ROOT` setting.
 * **staticfiles.E003**: The prefix ``<prefix>`` in the
   :setting:`STATICFILES_DIRS` setting must not end with a slash.
+* **staticfiles.W001**: The directory ``<directory>`` in the
+  :setting:`STATICFILES_DIRS` must exist.
```
Comment (felixxm) [inline docs/ref/checks.txt:873]:
```suggestion
  :setting:`STATICFILES_DIRS` doesn't exist.
```

File: docs/ref/checks.txt [hunk 2]
```diff
@@ -869,3 +869,5 @@ configured:
   contain the :setting:`STATIC_ROOT` setting.
 * **staticfiles.E003**: The prefix ``<prefix>`` in the
   :setting:`STATICFILES_DIRS` setting must not end with a slash.
+* **staticfiles.W001**: The directory ``<directory>`` in the
```
Comment (felixxm) [inline docs/ref/checks.txt:872]:
```suggestion
* **staticfiles.W004**: The directory ``<directory>`` in the
```

File: django/contrib/staticfiles/finders.py [hunk 3]
```diff
@@ -75,6 +75,9 @@ def check(self, **kwargs):
                 hint='Perhaps you forgot a trailing comma?',
                 id='staticfiles.E001',
             ))
+            if isinstance(settings.STATICFILES_DIRS, str):
+                # Iterating each character below would not be helpful.
+                return errors
```
Comment (felixxm) [inline django/contrib/staticfiles/finders.py:80]: This is a separate cleanup, so please move it do a separate commit.

File: django/contrib/staticfiles/finders.py [hunk 4]
```diff
@@ -90,6 +93,13 @@ def check(self, **kwargs):
                     'STATIC_ROOT setting.',
                     id='staticfiles.E002',
                 ))
+            elif not os.path.isdir(root):
+                errors.append(Warning(
+                    f'The directory {root} in the STATICFILES_DIRS setting '
+                    'does not exist.',
+                    hint='Perhaps the directory is not tracked in Git?',
```
Comment (felixxm) [inline django/contrib/staticfiles/finders.py:100]: I don't see much value in this hint, also `Git` is not the only VCS. Please remove it.

**Review by felixxm — COMMENTED** [review]
@jacobtylerwalls Thanks :+1:

File: django/contrib/staticfiles/finders.py (hunk 3 above)
Comment (tim-mccurrach) [reply django/contrib/staticfiles/finders.py:80]: Do we need to check that it's a string here? 

If it's not a list, a tuple, it seems unlikely to be some other iterable for which the further checks would be useful. As such - most likely an error will be raised when it's iterated over and the error message above won't even be displayed. (I'm imaging some people might use a Path instance).

This will then become a hard-to-diagnose problem for beginners, who will have a confusing exception rather than a useful error message from django. I wonder if for their sake it's worth always returning here.

File: django/contrib/staticfiles/finders.py (hunk 4 above)
Comment (jacobtylerwalls) [reply django/contrib/staticfiles/finders.py:100]: Yeah, that occurred to me also, but I guess I was trying to address the reporter's concern that beginners will still struggle without a hint. But that's what SO is for. I'm happy to remove it.

File: django/contrib/staticfiles/finders.py (hunk 3 above)
Comment (jacobtylerwalls) [reply django/contrib/staticfiles/finders.py:80]: Yeah if a `dict` or `set` falls through, it will just crash on `os.path.abspath()`, so I agree further checks are extremely unlikely to happen anyway. I'll just return here.

File: django/contrib/staticfiles/finders.py [hunk 5]
```diff
@@ -75,6 +75,7 @@ def check(self, **kwargs):
                 hint='Perhaps you forgot a trailing comma?',
                 id='staticfiles.E001',
             ))
+            return
```
Comment (tim-mccurrach) [inline django/contrib/staticfiles/finders.py:78]:
```suggestion
            return errors
```

Comment (jacobtylerwalls) [reply django/contrib/staticfiles/finders.py:78]: Thanks, haste makes waste!

**Review by felixxm — COMMENTED** [review]
@jacobtylerwalls Thanks for updates :+1:

I reorganized commits, added a small cleanup with using `CollectionTestCase` in `FindersCheckTests` tests, and pushed minor edits.

**Review by carltongibson — APPROVED** [review]
OK, yes, no problem, super, thanks all.

File: tests/staticfiles_tests/test_checks.py [hunk 6]
```diff
@@ -1,13 +1,18 @@
...
+from django.contrib.staticfiles.finders import BaseFinder, get_finder
+from django.core.checks import Error, Warning
+from django.test import override_settings
 
+from .cases import CollectionTestCase
+from .settings import TEST_ROOT
 
-class FindersCheckTests(SimpleTestCase):
+
+class FindersCheckTests(CollectionTestCase):
```
Comment (felixxm) [inline tests/staticfiles_tests/test_checks.py:14]: Using `CollectionTestCase` gives us existing directories in "static" settings.
